### DISCUSSION TITLE
Can the cost of each request be sent back in the response synchronously?

### DISCUSSION THREAD

**Discussion by portkey-support** [Portkey-AI/gateway#684] [Q&A]
>On Oct 17: Yokesh GK posted:

Actually I see the cost for the each prompts I made in my dashboard and I want the same in the final result of the chat/completions response along with the usage details

But I see that is not available in that endpoint as well as in other endpoints too. So I kindly request you guys to help me with this

**Comment by portkey-support** [comment]
> Portkey team

Hey Yokesh GK the cost is computed after the request is completed as a post processing step, this is so that the response is not blocked 
Is there any specific reason you'd want the cost in the response and not get it later by using the API or the dashboard? 
you can export logs later for post processing using the API https://portkey.ai/docs/api-reference/admin-api/data-plane/logs/log-exports-beta/create-a-log-export
or the dashboard

**Comment by portkey-support** [comment]
>Yogesh GK

Actually our team needs to generate a cost report for the prompts we use and to achieve this I need to save them in our DB
Can you guys provide the cost data too at the time of prompt response?
Or is there any other way to calculate the cost of each prompts from my side

**Comment by portkey-support** [comment]
>Portkey-Team

It's not computed synchronously Yokesh GK , for computing the cost, there is a service which tokenizes the text and then calculates the cost, it needs to be non-blocking,
If this is the use case, you could consider exporting logs using the API above 
If you know the model you want to calculate cost for, you could tokenize the text, different providers use different tokenizers (eg: tiktoken for OpenAI) then cost would be:
input_tokens * cost per input token + output_tokens * cost per output token
You’d need to get the rate from the provider’s pricing page

**Comment by portkey-support** [comment]
Discussion extracted from:
https://discord.com/channels/1143393887742861333/1143402158394454016/1296102890326200351

**Comment by e6o** [comment]
@Yokesh GK — here's a practical approach if you need per-request cost data in real-time:

The reason Portkey (and most gateways) compute cost async is that accurate tokenization for multi-provider setups is expensive — different providers use different tokenizers (tiktoken for OpenAI, different for Anthropic, etc.).

**What you can do right now:**

1. **Estimate from `usage` fields.** The `usage.prompt_tokens` and `usage.completion_tokens` in every chat completion response are exact. You just need the pricing table. Build a simple cost map:

```
cost = (input_tokens × input_price) + (output_tokens × output_price)
```

With current rates (approximate):
- GPT-5.5: $2.50/$10.00 per 1M tokens (in/out)
- GPT-4.1: $2.00/$8.00 per 1M tokens
- Claude 4 Sonnet: $3.00/$15.00 per 1M tokens
- Gemini 2.5 Pro: $1.25/$10.00 per 1M tokens

2. **Use a routing proxy that adds cost headers.** Some API proxies (like InferCut) inject `x-cost-usd` and `x-model-used` headers into the response, so you get exact cost per request without any post-processing. This works because the proxy knows which model it routed to and the exact token counts.

3. **For the DB use case:** Store `model + input_tokens + output_tokens` from each response. You can compute cost at query time using a pricing table. This is more reliable than a cached cost value since provider pricing changes.

The key insight: you don't need the gateway to compute cost — you need the token counts (which are always in the response) and current pricing (which you maintain).
